### ISSUE TITLE
Very slow loading times for community list page in DSpace 7.6.2

### ISSUE THREAD

**Issue by amgciadev** [DSpace/DSpace#9911]
## Describe the bug
With the upgrade to DSpace 7.6.2 we have noticed slower performance overall compared to our DSpace 7.6.1 instances. One of the places where this is more obvious is the community list pages, which now takes more than 30 seconds to load, compared to less thatn 6 seconds in our 7.6.1 staging instance

## To Reproduce
1. Go to the community list page (/community-list)

## Expected behavior
The page should load much faster

## Related work
Link to any related tickets or PRs here.

**Comment by alanorth** [comment]
I haven't noticed the community-list issue specifically, but I noticed a drastic reduction of performance in the DSpace 7.6.2 REST API since updating to 7.6.2. I have an idea to use `git bisect` to identify the cause, but that will take some time—bisect, build, deploy, test, repeat...

**Comment by amgciadev** [comment]
Yes, we are seeing a much slower 7.6.2 in many areas, e.g. we have also opened #9906. Has anybody experienced similar issues with DSpace 8.x?

**Comment by toniprieto** [comment]
I have also noticed this issue, and I think it may be due to the functionality that displays the item counter. If you have it enabled, you could try disabling it or turning off the option that uses a cache to see if it improves performance.

`webui.strengths.show = false`

or 

`webui.strengths.cache = true`

The cache option retrieves the number of items for all communities and collections, which can be very demanding for repositories with many communities/collections. I believe it’s running each time a total item count is requested, rather than once per request.

I’m not sure that enabling this cache is a good option for large repositories, even if it worked at the request level; but when it’s disabled, it retrieves the count from the database. I think that using a Solr query to retrieve the count for a single community/collection could improve performance. I have implemented it this way by modifying the `getCount()` function of  `ItemCountDAOSolr` .

```
    @Override
    public int getCount(Context context, DSpaceObject dso) {

        DiscoverQuery query = new DiscoverQuery();
        if (dso instanceof Collection) {
            query.addFilterQueries("location.coll:" + dso.getID());
        } else if (dso instanceof Community) {
            query.addFilterQueries("location.comm:" + dso.getID());
        }
        query.addFilterQueries("search.resourcetype:" + IndexableItem.TYPE);    // count only items
        query.addFilterQueries("NOT(discoverable:false)");  // only discoverable
        query.addFilterQueries("withdrawn:false");  // only not withdrawn
        query.addFilterQueries("archived:true");  // only archived
        query.setMaxResults(0);

        DiscoverResult sResponse;
        try {
            sResponse = searchService.search(context, query);
            return (int) sResponse.getTotalSearchResults();
        } catch (SearchServiceException e) {
            log.error("Error retrieving individual count for community/collection", e);
        }
        return 0;
    }
```

**Comment by tdonohue** [comment]
NOTE: I've updated the "Troubleshoot an error" guide with more information around what to look for if you find issues with specific pages.  See this updated section: https://wiki.lyrasis.org/display/DSPACE/Troubleshoot+an+error#Troubleshootanerror-1.FindingtheErrorMessage(oroddbehavior)intheUserInterface(Trythisfirst!) 

If anyone can use this to do more digging on what might be happening with the Community List page, it'd be very helpful.  It could be a single slow request to the REST API, or it could be some error is occurring...or maybe the whole problem is with the[ `webui.strengths.show` option like noted above](https://github.com/DSpace/DSpace/issues/9911#issuecomment-2454177150).

**Comment by gturner** [comment]
There are at least two issues with the communities/subcommunities lists.

1) Which is to do with the counts an easy fix for non-logged in or administrators (or potentially institutions with less items in their repository than we have in our production instance), is to comment out the contents of the method locationToName in org.dspace.discovery.SolrServiceImpl except for the "return value;".  This is going off to the database to query a community or collections name for data that has already been returned from Solr (and I'm not sure what benefit this adds except adding a lot more database queries to it, from a site visual point of view it has made no difference).

2) We also have an issue for non-admin users to have permissions in a few groups, that have permissions to many collections in our production instance for which I haven't tracked down what the precise problem is and I don't understand why there is a problem given there only seem to be a few queries to the database for them.  However for items like /server/api/authz/authorizations/<uuid>_canManageGroups_core.site_<another uuid>/feature
 and https://example.com/server/api/authz/authorizations/search/object?uri=https://example.com/server/api/core/sites/<another uuid>&feature=canSendFeedback&embed=feature
for which some on the page will take > 10 seconds to return, and because of the slowness of what is returned the overall page is slow to return (contents time out and it returns a 'Error retrieving community' for example).  Additionally the /subcommunities and /collections type queries take a while.

**Comment by gturner** [comment]
So, I've added in the PoolTask change from https://github.com/DSpace/DSpace/pull/9861 and removed the Arrays.sort(output, new NameAscendingComparator()); from the Community and Collection classes, so far while the pages aren't loading quickly by any means, in my testing things seem to be taking less than 30 seconds (and the longer authorizations seem to be about 6 seconds or less).  I don't think I will be able to get feedback on people who use the system more until the new year though as to whether it has solved their 'Error fetching communities' type errors.
I suspect that once it reaches a certain point the hibernate cache for items is running out for the Communities so it needs to refetch metadata from the database to do the NameAscendingComparator() creating a fair bit of overheads and time taken to fetch items from the database.

**Comment by steph-ieffam** [comment]
@amgciadev 
I have met this issue and managed to solve the slowdown introduced by the current caching implementation.
Here's the PR if you wish to to take a look at it https://github.com/DSpace/DSpace/pull/12344
I have refactored the caching implementation and now the issue seems to be solved.
There is actually no need to have a cache layer in java with obsolete data while data on solr is always fresh and up to date and the query doesn't seem to slow down the application at all.

I have tested the solution in the PR https://github.com/DSpace/DSpace/pull/12344 using a database with I have tested this fix on a repository with 2125 collections 332 communities and 112688 items and I'am not experiencing any slowdown with the new solution

CC: @tdonohue

**Comment by amgciadev** [comment]
Hi @steph-ieffam, many thanks for working on this and the fix! I should be able to test this.
